File: Leetcode/0097-interleaving-string/solution.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        """Time O(m*n), Space O(n)."""
        m, n = len(s1), len(s2)
        if m + n != len(s3): return False
        dp = [False] * (n + 1)
        dp[0] = True
        for j in range(1, n + 1):
            dp[j] = dp[j - 1] and s2[j - 1] == s3[j - 1]
        for i in range(1, m + 1):
            dp[0] = dp[0] and s1[i - 1] == s3[i - 1]
            for j in range(1, n + 1):
                dp[j] = (dp[j] and s1[i - 1] == s3[i + j - 1]) or \
                        (dp[j - 1] and s2[j - 1] == s3[i + j - 1])
        return dp[n]


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    cases = [
        ("Example 1", "aabcc", "dbbca", "aadbbcbcac", True),
        ("Example 2", "aabcc", "dbbca", "aadbbbaccc", False),
        ("All empty", "", "", "", True),
        ("s1 empty", "", "abc", "abc", True),
        ("s2 empty", "abc", "", "abc", True),
        ("Length mismatch", "a", "b", "abc", False),
        ("Same chars", "ab", "ba", "abba", True),
    ]
    for n, s1, s2, s3, exp in cases:
        got = sol.isInterleave(s1, s2, s3)
        if got == exp: print(f"PASS: {n}"); passed += 1
        else: print(f"FAIL: {n} got={got}"); failed += 1
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")

File: Leetcode/0097-interleaving-string/solution.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        int m = s1.length(), n = s2.length();
        if (m + n != s3.length()) return false;
        boolean[] dp = new boolean[n + 1];
        dp[0] = true;
        for (int j = 1; j <= n; j++) dp[j] = dp[j - 1] && s2.charAt(j - 1) == s3.charAt(j - 1);
        for (int i = 1; i <= m; i++) {
            dp[0] = dp[0] && s1.charAt(i - 1) == s3.charAt(i - 1);
            for (int j = 1; j <= n; j++) {
                dp[j] = (dp[j] && s1.charAt(i - 1) == s3.charAt(i + j - 1)) ||
                        (dp[j - 1] && s2.charAt(j - 1) == s3.charAt(i + j - 1));
            }
        }
        return dp[n];
    }

    static int passed = 0, failed = 0;
    public static void main(String[] args) {
        Solution sol = new Solution();
        Object[][] cases = {
            {"Example 1", "aabcc", "dbbca", "aadbbcbcac", true},
            {"Example 2", "aabcc", "dbbca", "aadbbbaccc", false},
            {"All empty", "", "", "", true},
            {"s1 empty", "", "abc", "abc", true},
            {"s2 empty", "abc", "", "abc", true},
            {"Length mismatch", "a", "b", "abc", false},
            {"Same chars", "ab", "ba", "abba", true}
        };
        for (Object[] c : cases) {
            boolean got = sol.isInterleave((String) c[1], (String) c[2], (String) c[3]);
            if (got == (boolean) c[4]) { System.out.println("PASS: " + c[0]); passed++; }
            else { System.out.println("FAIL: " + c[0] + " got=" + got); failed++; }
        }
        System.out.printf("%nResults: %d passed, %d failed, %d total%n", passed, failed, passed + failed);
    }
}
